doc/decisions/1_in_discussion/internal_cache.md

```diff
@@ -0,0 +1,350 @@
+assert (keyString(key) != keyString(key_dup));
+assert (keySetName (key_dup, "dir:/valid") == -1); // must fail, as we have a COW key
+assert (keyName(key) == keyName(key_dup)); // stays always valid
+```
+
+This is already implemented for the MMAP cache, so the implementation should be straightforward:
+Do the same COW duplications as done for MMAP but with a different flag.
+
+@kodebach wrote:
+
+> What I wanted to say is that mmap already does COW, so we can reuse the code and probably the flag.
+> If there is some code that is only needed for mmap and not COW, we could make mmap set two flags one for mmap and one for the general COW code.
+
+For the metadata, however, also COW KeySets might be needed (at least with the current API).
+Example:
+
+```c
+keyCopy (cow, key, ELEKTRA_CP_COW);
+KeySet * cowMeta = keyMeta (cow);
+ksAppendKey (cowMeta, keyNew ("meta:/whatever", KEY_VALUE, "abc", KEY_END));
+ksRemoveByName (cowMeta, "meta:/type");
+```
+
+**Pros:**
+
+- Elektra doesn't require MMAP
+
+**Cons:**
+
+- Lifetime of a copied COW key MUST be less than the key it was copied from.
+  We can not track how many keys point to the same data and name this way, so we can only free data and name if the key does not have the COW flag.
+  If the original key gets deleted, using a COW key that points to the same data and name will lead to corrupt data.
+  The same is true for updating values of the original key.
+  
+  This is only problematic if we want to use COW for keys outside of `KDB`.
+  If it is only for use within `KDB`, especially for usage as internal cache and in change tracking, we could always guarantee that the original keys are going to last as long as the `KDB` instance.
+  However, we need to document for the users of Elektra that keys returned from `kdbGet` are only valid until `kdbClose`.
+  If they want to continue using them afterwards, they'd have to deep copy them.
+  
+  Triggering the delete problem:
+  ```c
+  Key * originalKey;
+  Key * copiedKey;
+  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
+  
+  assert (keyString (copiedKey) == keyString (originalKey));
+  assert (keyName (copiedKey) == keyName (originalKey));
+
+  keyDel (originalKey);
+  
+  keyString (copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to freed memory
+  keyName (copiedKey);   // Error! Original name has been deleted.
+  ```
+ 
+  Triggering the update problem:
+  ```c
+  Key * originalKey;
+  Key * copiedKey;
+  keyCopy (copiedKey, originalKey, ELEKTRA_CP_COW);
+  
+  assert (keyString(copiedKey) == keyString(originalKey));
+  keySetString (originalKey, "new value!");
+  
+  keyString(copiedKey); // Error! Original value has been deleted. Pointer to data in copiedKey points to potentially freed memory
+  ```
+
+  The same problems in principle exist for `mmapstorage`, however this is just a storage plugin and the user does not have any API to create COW keys themselves.
```

Comment:
Not only principle, I think @kodebach is right that `kdbClose` will unmap everything, so the KeySet afterwards is broken.

I think this is currently also not properly documented, so we definitely must do something about it.

```suggestion
  The same problems exist for `mmapstorage` where `kdbClose` frees (munmap) the KeySet.
```

Reply:
As stated elsewhere, I checked and `munmap` actually happens in `kdbSet` for `mmapstorage`. Note: I have not checked when this would be called for the cache.

Reply:
[217dff4](https://github.com/ElektraInitiative/libelektra/pull/4641/commits/217dff470ccf40b28993628aaf6d91eabdf2d545)

Reply:
@mpranj can you help out here?

(That `kdbSet` does it might be true but is totally uninteresting for the API user.)

Reply:
> That `kdbSet` does it might be true but is totally uninteresting for the API user

It's not phrased perfectly. But it should tell the API user, that the keyset is only valid until `kdbSet` is called.

In any case, this document is not addressed to API users, so it doesn't matter if it is interesting to API users or not, if it is interesting to the audience of the decision. Which it is, because the only interesting bit here is where `munmap` happens.